testing/monkey/monkey_dev.c

```diff
@@ -0,0 +1,327 @@
+
+static bool button_read(int fd, FAR uint32_t *value)
+{
+  btn_buttonset_t buttonset;
+
+  int ret = read(fd, &buttonset, sizeof(buttonset));
+  if (ret < 0)
+    {
+      return false;
+    }
+
+  *value = buttonset;
+  return true;
+}
+
+/****************************************************************************
+ * Name: monkey_dev_create
+ ****************************************************************************/
+
+FAR struct monkey_dev_s *monkey_dev_create(FAR const char *dev_path,
+                                           enum monkey_dev_type_e type)
+{
+  FAR struct monkey_dev_s *dev;
+  int oflag;
+  int fd;
+
+  MONKEY_ASSERT_NULL(dev_path);
+
+  dev = malloc(sizeof(struct monkey_dev_s));
+  MONKEY_ASSERT_NULL(dev);
+  memset(dev, 0, sizeof(struct monkey_dev_s));
+
+  if (MONKEY_IS_UINPUT_TYPE(type))
+    {
+      oflag = O_RDWR | O_NONBLOCK;
+    }
+  else
+    {
+      oflag = O_RDONLY | O_NONBLOCK;
+    }
+
+  fd = open(dev_path, oflag);
+  if (fd < 0)
+    {
+      MONKEY_LOG_ERROR("open %s failed: %d", dev_path, errno);
+      goto failed;
+    }
+
+  dev->type = type;
+  dev->fd = fd;
+
+  MONKEY_LOG_NOTICE("open %s success, fd = %d, type: %s",
+                    dev_path, fd, monkey_dev_type2name(type));
+
+  return dev;
+
+failed:
+  if (fd > 0)
+    {
+      close(fd);
+    }
+
+  if (dev)
+    {
+      free(dev);
+    }
+
+  return NULL;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: monkey_dev_delete
+ ****************************************************************************/
+
+void monkey_dev_delete(FAR struct monkey_dev_s *dev)
+{
+  MONKEY_ASSERT_NULL(dev);
+
+  if (dev->fd > 0)
+    {
+      /* Reset input state */
+
+      struct monkey_dev_state_s state;
+      memset(&state, 0, sizeof(state));
+      state.type = dev->type;
+      monkey_dev_set_state(dev, &state);
+
+      MONKEY_LOG_NOTICE("close fd: %d", dev->fd);
+      close(dev->fd);
+      dev->fd = -1;
+    }
+
+  free(dev);
+}
+
+/****************************************************************************
+ * Name: monkey_dev_set_state
+ ****************************************************************************/
+
+void monkey_dev_set_state(FAR struct monkey_dev_s *dev,
+                           FAR const struct monkey_dev_state_s *state)
+{
+  MONKEY_ASSERT_NULL(dev);
+
+  switch (MONKEY_GET_DEV_TYPE(dev->type))
+    {
+    case MONKEY_DEV_TYPE_TOUCH:
+      utouch_write(dev->fd,
+                   state->data.touch.x,
+                   state->data.touch.y,
+                   state->data.touch.is_pressed);
+      break;
+
+    case MONKEY_DEV_TYPE_BUTTON:
+      ubutton_write(dev->fd, state->data.button.value);
+      break;
+
+    default:
+      break;
```

Comment:
I think 2 more spaces are needed for case block

Reply:
done